text/3417-dropck-eyepatch-v3.md

```diff
@@ -0,0 +1,250 @@
+- Feature Name: (`dropck_eyepatch_v3`)
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Cleanup the rules for implicit drops by adding an argument for `#[may_dangle]` on type
+parameters: `#[may_dangle(droppable)]` and `#[may_dangle(must_not_use)]`. Change `PhantomData`
+to get completely ignored by dropck as its current behavior is confusing and inconsistent.
+
+# Motivation
+[motivation]: #motivation
+
+`PhantomData` is `Copy` but still adds outlives requirements when dropped as a part of
+a larger value. This behavior is inconsistent and results in "spooky-dropck-at-a-distance" 
+even without `#[may_dangle]`:
+
+```rust
+use std::marker::PhantomData;
+struct PrintOnDrop<'a>(&'a str); // requires `'a` to be live on drop.
```

Comment:
e.g., Consider first a type `PrintOnDrop` which holds a `&str` reference and has a `Drop` impl. This impl will access data from the reference when it is dropped. In general the compiler assumes that every drop impl will access all the data from the structs fields.